File: two_pointers/3349.cpp
// Brute force
// Tc = O(n + k)
// Sc = O(1)
class Solution {
public:
    bool isIncreasing(vector<int>& nums, int s, int e) {
        for (int i = s + 1; i < e; i++) {
            if (nums[i] <= nums[i - 1]) { // not increasing
                return false;
            }
        }
        return true;
    }
    bool hasIncreasingSubarrays(vector<int>& nums, int k) {
        int n = nums.size();
        for (int start = 0; start + 2 * k <= n; start++) {
            bool first = isIncreasing(nums, start, start + k);
            bool second = isIncreasing(nums, start + k, start + k * 2);
            if (first && second) {
                return true;
            }
        }
        return false;
    }
};

// Optimal
class Solution {
public:
    bool hasIncreasingSubarrays(vector<int>& nums, int k) {
        int n = nums.size();
        int currRun = 1;
        int prevRun = 0;
        for (int i = 1; i < n; i++) {
            if (nums[i] > nums[i - 1]) {
                currRun++;
            } else {
                // increasing order break
                prevRun = currRun;
                currRun = 1;
            }
            if (currRun >= 2 * k) {
                // can accomodate two sub array in it length >= k
                return true;
            }
            if (min(currRun, prevRun) >= k) {
                return true;
            }
        }
        return false;
    }
};

File: two_pointers/3349.py
class Solution:
    def isIncreasing(self, nums: List[int], s: int, e: int) -> bool:
        for i in range(s + 1, e):
            if nums[i] <= nums[i - 1]:
                return False
        return True

    def hasIncreasingSubarrays(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        for start in range(n - 2 * k + 1):
            first = self.isIncreasing(nums, start, start + k)
            second = self.isIncreasing(nums, start + k, start + 2 * k)
            if first and second:
                return True
        return False
    

class Solution:
    def hasIncreasingSubarrays(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        currRun = 1
        prevRun = 0
        for i in range(1, n):
            if nums[i] > nums[i-1]:
                currRun += 1
            else:
                prevRun = currRun
                currRun = 1
            if currRun >= 2*k:
                return True
            elif min(currRun, prevRun) >= k:
                return True
        return False
